src/cli/context.js

```diff
@@ -29,17 +32,66 @@ class Context {
   async init() {
     const { rawArguments, logger } = this;
 
-    const { plugins } = parseArgvWithoutPlugins(rawArguments, logger, [
+    const {
+      plugins,
+      ignorePath,
+      withNodeModules,
+      config,
+      editorconfig,
+      _: filePatterns,
+    } = parseArgvWithoutPlugins(rawArguments, logger, [
       "plugin",
+      "ignore-path",
+      "with-node-modules",
+      "config",
+      "editorconfig",
+      "_",
     ]);
 
-    await this.pushContextPlugins(plugins);
+    const pluginsFromConfigFile = await this.#loadPluginsFromConfigFile(
+      filePatterns,
+      {
+        config,
+        editorconfig,
+        ignorePath,
+        withNodeModules,
+      },
+    );
+
+    await this.pushContextPlugins([...plugins, ...pluginsFromConfigFile]);
 
     const argv = parseArgv(rawArguments, this.detailedOptions, logger);
     this.argv = argv;
     this.filePatterns = argv._;
   }
 
+  async #loadPluginsFromConfigFile(filePatterns, argv) {
+    const { config, editorconfig, ignorePath, withNodeModules } = argv;
+    const isIgnored = await createIsIgnoredFromContextOrDie({
+      logger: this.logger,
+      argv: { ignorePath, withNodeModules },
+    });
+    const plugins = [];
+    const cwd = process.cwd();
+    for (const filePattern of filePatterns) {
+      const absolutePath = path.resolve(cwd, filePattern);
+      if (isIgnored(absolutePath)) {
+        continue;
+      }
+      const options = await getOptionsOrDie(
+        { logger: this.logger, argv: { config, editorconfig } },
+        absolutePath,
+        /* getAllOverridesPlugins */ true,
+      );
+      if (Array.isArray(options?.plugins)) {
+        for (const plugin of options.plugins) {
+          plugins.push(plugin);
+        }
+      }
+    }
+    return plugins;
```

Comment:
I don't quite understand this part. Should we only need collect plugins and load them to get `filenames` and `extensions` to use here? https://github.com/prettier/prettier/blob/b37a120ab6459f25218b411baab66ee73de02af9/src/cli/expand-patterns.js#L135

Reply:
Yes

Reply:
Then why we need this ignore check here?

Reply:
Ah, I got it file path is needed to resolve config, right?

Reply:
Even worth case.

```
-dir (.prettierrc (empty) and 1.foo file)
-dir/subdir (.prettier (with prettier-plugin-foo)  and 2.foo file)

prettier dir dir/subdir
```

We'll glob `1.foo` and `2.foo`, and error will throw when formatting `1.foo`

Reply:
Another problem.

```
# .prettierignore
1
!1/2
```

```
prettier 1
```

`1/` is ignored, but `1/2` is unignored, we won't read `.prettierrc` in the dir, we won't glob files in `1/2/*.foo` either.